src/builtins.cpp

```diff
@@ -321,957 +235,346 @@ static void lUpdateIntrinsicsAttributes(llvm::Module *module) {
-        // separated from the others and it is not done by default
-        lUpdateIntrinsicsAttributes(bcModule);
+llvm::Constant *lFuncAsConstInt8Ptr(llvm::Module &M, const char *name) {
+    llvm::LLVMContext &Context = M.getContext();
+    llvm::Function *F = M.getFunction(name);
+    if (F) {
+        return llvm::ConstantExpr::getBitCast(F, llvm::Type::getInt8PtrTy(Context));
     }
+    return nullptr;
+}
+
+void lRemoveUnused(llvm::Module *M) {
+    llvm::FunctionAnalysisManager FAM;
+    llvm::ModuleAnalysisManager MAM;
+    llvm::ModulePassManager PM;
+    llvm::PassBuilder PB;
+    PB.registerModuleAnalyses(MAM);
+    PM.addPass(llvm::GlobalDCEPass());
+    PM.run(*M, MAM);
+}
 
-    // A hack to move over declaration, which have no definition.
-    // New linker is kind of smart and think it knows better what to do, so
-    // it removes unused declarations without definitions.
-    // This trick should be legal, as both modules use the same LLVMContext.
-    for (llvm::Function &f : *bcModule) {
-        if (f.isDeclaration()) {
-            // Declarations with uses will be moved by Linker.
-            if (f.getNumUses() > 0)
-                continue;
-            module->getOrInsertFunction(f.getName(), f.getFunctionType(), f.getAttributes());
+void lRemoveUnusedPersistentFuntions(llvm::Module *M) {
+    // TODO! make it clean. The idea here is to preserve all persistent
+    // functions in the first call of addPersistentToLLVMUsed then
+    // remove dead code. After that, manually remove llvm.compiler.used
+    // as well as uses of that functions that referenced by bitcats
+    // constant exprs. Then we rerun addPersistentToLLVMUsed, on the
+    // second run it preserves only subset of functions that are really
+    // used (or can be used) in module.
+    // M->print(llvm::errs(), nullptr);
+    llvm::GlobalVariable *llvmUsed = M->getNamedGlobal("llvm.compiler.used");
+    if (llvmUsed) {
+        std::unordered_map<llvm::Function *, int> usedFunctions;
+        llvm::ConstantArray *initList = llvm::cast<llvm::ConstantArray>(llvmUsed->getInitializer());
+        for (unsigned i = 0; i < initList->getNumOperands(); i++) {
+            auto *C = initList->getOperand(i);
+            // C->dump();
+            llvm::ConstantExpr *CE = llvm::dyn_cast<llvm::ConstantExpr>(C);
+            // CE->dump();
+            // Bitcast as ConstExpr when opaque pointer is not used, otherwise C is just an opaque pointer.
+            llvm::Value *val = CE ? CE->getOperand(0) : C;
+            Assert(val);
+            // val->dump();
+            if (val->getNumUses() > 1) {
+                Assert(llvm::isa<llvm::Function>(val));
+                usedFunctions[llvm::cast<llvm::Function>(val)] = 1;
+            }
+        }
+
+        // Extend the list of preserved functions with the functions from corresponding persistent groups.
+        std::unordered_map<const builtin::PersistentGroup *, int> usedPersistentGroups;
```

Comment:
I'm not quite understand why unordered_map is needed. Maybe unordered_set would be sufficient?

Reply:
No specific reason for that, unordered set will suit.